iterated-prisoners-dilemma/aniket/evolution.py

```diff
@@ -0,0 +1,143 @@
+import random
+import sys
+
+def assign_score(bit_a, bit_b):
+    """assign score corresponding to the set of moves
+       chosen by the two players.
+       bit value of 1 --> cooperation
+       bit value of 2 --> defection"""
+
+    if (bit_a,bit_b)==('1','1'):
+        return (3,3)
+    elif (bit_a,bit_b)==('1','0'):
+        return (0,5)
+    elif (bit_a,bit_b)==('0','1'):
+        return (5,0)
+    else:
+        return (1,1)  
+
+def get_strategy(encoding):
+    """return the index in the strategy 
+       corresponding to the initial memory of the
+       3 games"""
+
+    return int(encoding, 2)
+
+def play_game(strategy_a, strategy_b):
+    """we have a total of 100 games between two strategies"""
+    
+    score_a = score_b = 0
+    encoding_a = strategy_a[64:70]
+    encoding_b = strategy_b[64:70]
+    for _ in range (0, 100):
+        move_a = strategy_a[get_strategy(encoding_a)]
+        move_b = strategy_b[get_strategy(encoding_b)]
+        a, b = assign_score(move_a, move_b)
+        score_a += a
+        score_b += b
+        encoding_a = encoding_a[2:]+ move_a + move_b
+        encoding_b = encoding_b[2:]+ move_b + move_a
+    #averaging the scores over 100 games        
+    score_a/=100
+    score_b/=100        
+    return (score_a, score_b)    
+                
+
+def fitness_calculation(strategies, fitness, num_organisms):
+    """each strategy plays with all other (num_organisms-1) strategies 
+       and with itself i.e. a total of num_organisms games"""
+
+    for i in range(0, num_organisms):
+        for j in range(i, num_organisms):
+            x, y = play_game(strategies[i], strategies[j])
+            fitness[i]+=x
+            fitness[j]+=y
+    #averaging the fitness values over num_organisms strategies            
+    fitness = [f/num_organisms for f in fitness]
+    return fitness
+
+def fitness_proportionate_selection(strategies, fitness, num_organisms):
+    """choosing parent strategies on the basis of roulette 
+       wheel selection, and we perform this process for 'n'
+       number of times, where 
+       n --> number of strategies, here num_organisms."""
+
+    max_val = sum(fitness)
+    parents = []
+    for _ in range(num_organisms):
+        temp_sum = 0
+        rand_val = random.uniform(0, max_val)
+        for j in range(num_organisms):
+            temp_sum += fitness[j]
+            if temp_sum>=rand_val:
```

Comment:
Assume that `j = 0` and `fitness[j] = 2` whereas `max_val = 100`. Then there's a 98% chance that `rand_val` will be greater than `2`. But this should not have been the case for such a low fitness individual. Maybe you meant to write `temp_sum <= rand_val`.